

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>selene_utils2 module &mdash; Orca 0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="orca_modules module" href="orca_modules.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Orca
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="orca_predict.html">orca_predict module</a></li>
<li class="toctree-l1"><a class="reference internal" href="orca_utils.html">orca_utils module</a></li>
<li class="toctree-l1"><a class="reference internal" href="orca_models.html">orca_models module</a></li>
<li class="toctree-l1"><a class="reference internal" href="orca_modules.html">orca_modules module</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">selene_utils2 module</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Orca</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>selene_utils2 module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/selene_utils2.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-selene_utils2">
<span id="selene-utils2-module"></span><h1>selene_utils2 module<a class="headerlink" href="#module-selene_utils2" title="Permalink to this headline">¶</a></h1>
<p>This module provides the selene-based utilities for training and using
Orca sequence models for multiscale genome interaction prediction. This
module contains code from selene.</p>
<dl class="class">
<dt id="selene_utils2.Genomic2DFeatures">
<em class="property">class </em><code class="sig-prename descclassname">selene_utils2.</code><code class="sig-name descname">Genomic2DFeatures</code><span class="sig-paren">(</span><em class="sig-param">input_paths</em>, <em class="sig-param">features</em>, <em class="sig-param">shape</em>, <em class="sig-param">cg=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#Genomic2DFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.Genomic2DFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.targets.target.Target</span></code></p>
<p>Stores one or multple datasets of Hi-C style 2D data in cooler format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_paths</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – List of paths to the Cooler datasets. For mcool files,
the path should include the resolution. Please refer to
cooler.Cooler documentation for support of mcool files.</p></li>
<li><p><strong>features</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – The list of dataset names that should match the <cite>input_path</cite>.</p></li>
<li><p><strong>shape</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>)</em>) – The shape of the output array (# of bins by # of bins).</p></li>
<li><p><strong>cg</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – If <cite>yes</cite>, adpative coarse-graining is applied to the output.</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="selene_utils2.Genomic2DFeatures.data">
<code class="sig-name descname">data</code><a class="headerlink" href="#selene_utils2.Genomic2DFeatures.data" title="Permalink to this definition">¶</a></dt>
<dd><p>The list of Cooler objects for the cooler files.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(cooler.Cooler)</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.Genomic2DFeatures.n_features">
<code class="sig-name descname">n_features</code><a class="headerlink" href="#selene_utils2.Genomic2DFeatures.n_features" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of cooler files.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.Genomic2DFeatures.feature_index_dict">
<code class="sig-name descname">feature_index_dict</code><a class="headerlink" href="#selene_utils2.Genomic2DFeatures.feature_index_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping feature names (<cite>str</cite>) to indices (<cite>int</cite>),
where the index is the position of the feature in <cite>features</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.Genomic2DFeatures.shape">
<code class="sig-name descname">shape</code><a class="headerlink" href="#selene_utils2.Genomic2DFeatures.shape" title="Permalink to this definition">¶</a></dt>
<dd><p>The shape of the output array (# of bins by # of bins).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.Genomic2DFeatures.cg">
<code class="sig-name descname">cg</code><a class="headerlink" href="#selene_utils2.Genomic2DFeatures.cg" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether adpative coarse-graining is applied to the output.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_utils2.Genomic2DFeatures.get_feature_data">
<code class="sig-name descname">get_feature_data</code><span class="sig-paren">(</span><em class="sig-param">chrom</em>, <em class="sig-param">start</em>, <em class="sig-param">end</em>, <em class="sig-param">chrom2=None</em>, <em class="sig-param">start2=None</em>, <em class="sig-param">end2=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#Genomic2DFeatures.get_feature_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.Genomic2DFeatures.get_feature_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve the feature data for some coordinate.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="selene_utils2.MemmapGenome">
<em class="property">class </em><code class="sig-prename descclassname">selene_utils2.</code><code class="sig-name descname">MemmapGenome</code><span class="sig-paren">(</span><em class="sig-param">input_path</em>, <em class="sig-param">blacklist_regions=None</em>, <em class="sig-param">bases_order=None</em>, <em class="sig-param">init_unpickable=False</em>, <em class="sig-param">memmapfile=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#MemmapGenome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.MemmapGenome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.sequences.genome.Genome</span></code></p>
<p>Memmapped version of selene.sequence.Genome. Faster for sequence
retrieval by storing all precomputed one-hot encodings in a memmapped
file (~40G for human genome).</p>
<p>The memmapfile can be an exisiting memmapped file or a path where you
want to create the memmapfile. If the specified memmapfile does not
exist, it will be created the first time you call any method of
MemmapGenome or if MemmapGenome is initialized with <cite>init_unpickable=True</cite>.
Therefore the first call will take some time for the
creation of memmapfile if it does not exist. Also,  if
memmapfile has not been created, be careful not to run multiple
instances of MemmapGenome in parallel (such as with Dataloader),
because as each process will try to create the file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Path to an indexed FASTA file, that is, a <cite>*.fasta</cite> file with
a corresponding <cite>*.fai</cite> file in the same directory. This file
should contain the target organism’s genome sequence.</p></li>
<li><p><strong>init_unpickleable</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – Default is False. If False, delay part of initialization code
to executed only when a relevant method is called. This enables
the object to be pickled after instantiation. <cite>init_unpickleable</cite> should
be <cite>False</cite> when used when multi-processing is needed e.g. DataLoader.</p></li>
<li><p><strong>memmapfile</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a><em>, </em><em>optional</em>) – Specify the numpy.memmap file for storing the encoding
of the genome. If memmapfile does not exist, it will be
created when the encoding is requested for the first time.</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="selene_utils2.MemmapGenome.genome">
<code class="sig-name descname">genome</code><a class="headerlink" href="#selene_utils2.MemmapGenome.genome" title="Permalink to this definition">¶</a></dt>
<dd><p>The FASTA file containing the genome sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pyfaidx.Fasta</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.MemmapGenome.chrs">
<code class="sig-name descname">chrs</code><a class="headerlink" href="#selene_utils2.MemmapGenome.chrs" title="Permalink to this definition">¶</a></dt>
<dd><p>The list of chromosome names.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.MemmapGenome.len_chrs">
<code class="sig-name descname">len_chrs</code><a class="headerlink" href="#selene_utils2.MemmapGenome.len_chrs" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping the names of each chromosome in the file to
the length of said chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_utils2.MemmapGenome.get_encoding_from_coords">
<code class="sig-name descname">get_encoding_from_coords</code><span class="sig-paren">(</span><em class="sig-param">chrom</em>, <em class="sig-param">start</em>, <em class="sig-param">end</em>, <em class="sig-param">strand='+'</em>, <em class="sig-param">pad=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#MemmapGenome.get_encoding_from_coords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.MemmapGenome.get_encoding_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the one-hot encoding of the genomic sequence at the
queried coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The name of the chromosome or region, e.g. “chr1”.</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The 0-based start coordinate of the first position in the
sequence.</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – One past the 0-based last position in the sequence.</p></li>
<li><p><strong>strand</strong> (<em>{'+'</em><em>, </em><em>'-'</em><em>, </em><em>'.'}</em><em>, </em><em>optional</em>) – Default is ‘+’. The strand the sequence is located on. ‘.’ is
treated as ‘+’.</p></li>
<li><p><strong>pad</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – Default is <cite>False</cite>. Pad the output sequence with ‘N’ if <cite>start</cite>
and/or <cite>end</cite> are out of bounds to return a sequence of length
<cite>end - start</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The <span class="math notranslate nohighlight">\(L \times 4\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L = end - start\)</span>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, dtype=numpy.float32</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#AssertionError" title="(in Python v3.9)"><strong>AssertionError</strong></a> – If it cannot retrieve encoding that matches the length <cite>L = end - start</cite>
    such as when end &gt; chromosome length and pad=False</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_utils2.MemmapGenome.get_encoding_from_coords_check_unk">
<code class="sig-name descname">get_encoding_from_coords_check_unk</code><span class="sig-paren">(</span><em class="sig-param">chrom</em>, <em class="sig-param">start</em>, <em class="sig-param">end</em>, <em class="sig-param">strand='+'</em>, <em class="sig-param">pad=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#MemmapGenome.get_encoding_from_coords_check_unk"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.MemmapGenome.get_encoding_from_coords_check_unk" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the one-hot encoding of the genomic sequence at the
queried coordinates and check whether the sequence contains
unknown base(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The name of the chromosome or region, e.g. “chr1”.</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The 0-based start coordinate of the first position in the
sequence.</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – One past the 0-based last position in the sequence.</p></li>
<li><p><strong>strand</strong> (<em>{'+'</em><em>, </em><em>'-'</em><em>, </em><em>'.'}</em><em>, </em><em>optional</em>) – Default is ‘+’. The strand the sequence is located on. ‘.’ is
treated as ‘+’.</p></li>
<li><p><strong>pad</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – Default is <cite>False</cite>. Pad the output sequence with ‘N’ if <cite>start</cite>
and/or <cite>end</cite> are out of bounds to return a sequence of length
<cite>end - start</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><cite>tuple[0]</cite> is the <span class="math notranslate nohighlight">\(L \times 4\)</span> encoding of the sequence, where</p></li>
</ul>
<p><span class="math notranslate nohighlight">\(L = end - start\)</span>.
<cite>L</cite> = 0 for the NumPy array returned.
* <cite>tuple[1]</cite> is the boolean value that indicates whether the
sequence contains any unknown base(s) specified in self.UNK_BASE</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a>(<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#AssertionError" title="(in Python v3.9)"><strong>AssertionError</strong></a> – If it cannot retrieve encoding that matches the length <cite>L = end - start</cite>
    such as when end &gt; chromosome length and pad=False</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_utils2.MemmapGenome.init">
<code class="sig-name descname">init</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#MemmapGenome.init"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.MemmapGenome.init" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="selene_utils2.MultibinGenomicFeatures">
<em class="property">class </em><code class="sig-prename descclassname">selene_utils2.</code><code class="sig-name descname">MultibinGenomicFeatures</code><span class="sig-paren">(</span><em class="sig-param">input_path</em>, <em class="sig-param">features</em>, <em class="sig-param">bin_size</em>, <em class="sig-param">step_size</em>, <em class="sig-param">shape</em>, <em class="sig-param">mode='center'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#MultibinGenomicFeatures"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.targets.target.Target</span></code></p>
<p>Multibin version of selene.targets.GenomicFeatures
Stores the dataset specifying features for genomic regions.
Accepts a <cite>*.bed</cite> file with the following columns,
in order:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">feature</span><span class="p">]</span>
</pre></div>
</div>
<p><cite>start</cite> and <cite>end</cite> is 0-based as in bed file format.</p>
<p>Note that unlike selene_sdk.targets.GenomicFeatures which queries
the tabix data file out-of-core, MultibinGenomicFeatures requires
more memory as it loads the entire bed file in memory as a pyranges
table for higher query speed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – Path to the bed file.</p></li>
<li><p><strong>features</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – The non-redundant list of genomic features names. The output array
will have the same feature order as specified in this list.</p></li>
<li><p><strong>bin_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The length of the bin(s) in which we check for features</p></li>
<li><p><strong>step_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The interval between two adjacent bins.</p></li>
<li><p><strong>shape</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>)</em>) – The shape of the output array (n_features by n_bins).</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>, </em><em>optional</em>) – For <cite>mode==’any’</cite>, any overlap will get 1, and no overlap will get 0.
For <cite>mode==’center’, only overlap with the center basepair of each bin
will get 1, otherwise 0.
For `mode==’proportion’</cite>, the proportion of overlap will be returned.</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="selene_utils2.MultibinGenomicFeatures.data">
<code class="sig-name descname">data</code><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.data" title="Permalink to this definition">¶</a></dt>
<dd><p>The data stored in PyRanges object.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pyranges.PyRanges</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.MultibinGenomicFeatures.n_features">
<code class="sig-name descname">n_features</code><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.n_features" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of distinct features.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.MultibinGenomicFeatures.feature_index_dict">
<code class="sig-name descname">feature_index_dict</code><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.feature_index_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping feature names (<cite>str</cite>) to indices (<cite>int</cite>),
where the index is the position of the feature in <cite>features</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.MultibinGenomicFeatures.index_feature_dict">
<code class="sig-name descname">index_feature_dict</code><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.index_feature_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping indices (<cite>int</cite>) to feature names (<cite>str</cite>),
where the index is the position of the feature in the input
features.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.MultibinGenomicFeatures.bin_size">
<code class="sig-name descname">bin_size</code><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.bin_size" title="Permalink to this definition">¶</a></dt>
<dd><p>The length of the bin(s) in which we check for features</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.MultibinGenomicFeatures.step_size">
<code class="sig-name descname">step_size</code><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.step_size" title="Permalink to this definition">¶</a></dt>
<dd><p>The interval between two adjacent bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.MultibinGenomicFeatures.shape">
<code class="sig-name descname">shape</code><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.shape" title="Permalink to this definition">¶</a></dt>
<dd><p>The shape of the output array (n_features by n_bins).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.MultibinGenomicFeatures.mode">
<code class="sig-name descname">mode</code><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.mode" title="Permalink to this definition">¶</a></dt>
<dd><ul class="simple">
<li><p>For <cite>mode==’any’</cite>, any overlap will get assigned 1, and no overlap will</p></li>
</ul>
<p>get assigned 0.
- For <cite>mode==’center’, only overlap with the center basepair of each bin
will get assigned 1, otherwise assigned 0.
- For `mode==’proportion’</cite>, the proportion of overlap will be assigned.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_utils2.MultibinGenomicFeatures.get_feature_data">
<code class="sig-name descname">get_feature_data</code><span class="sig-paren">(</span><em class="sig-param">chrom</em>, <em class="sig-param">start</em>, <em class="sig-param">end</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#MultibinGenomicFeatures.get_feature_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.get_feature_data" title="Permalink to this definition">¶</a></dt>
<dd><p>For a genomic region specified, return a <cite>number of features</cite>
by <cite>number of bins</cite> array for overlap of each genomic bin and
each feature. How the overlap is quantified depends on the
<cite>mode</cite> attribute specified during initialization.</p>
<p>For <cite>mode==’any’</cite>, any overlap will get assigned 1, and no overlap will
get assigned 0.
For <cite>mode==’center’, only overlap with the center basepair of each bin
will get assigned 1, otherwise assigned 0.
For `mode==’proportion’</cite>, the proportion of overlap will be assigned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The name of the region (e.g. ‘1’, ‘2’, …, ‘X’, ‘Y’).</p></li>
<li><p><strong>start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The 0-based first position in the region.</p></li>
<li><p><strong>end</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – One past the 0-based last position in the region.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><span class="math notranslate nohighlight">\(L \times N\)</span> array, where <span class="math notranslate nohighlight">\(L = \)</span>
and <span class="math notranslate nohighlight">\(N =\)</span> <cite>self.n_features</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_utils2.MultibinGenomicFeatures.init">
<code class="sig-name descname">init</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#MultibinGenomicFeatures.init"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.MultibinGenomicFeatures.init" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="selene_utils2.RandomPositionsSamplerHiC">
<em class="property">class </em><code class="sig-prename descclassname">selene_utils2.</code><code class="sig-name descname">RandomPositionsSamplerHiC</code><span class="sig-paren">(</span><em class="sig-param">reference_sequence, target, features, target_1d=None, background_cis_file=None, background_trans_file=None, seed=436, validation_holdout=['chr6', 'chr7'], test_holdout=['chr8', 'chr9'], sequence_length=1000000, max_seg_length=None, length_schedule=None, position_resolution=1, random_shift=0, random_strand=True, cross_chromosome=True, permute_segments=False, mode='train'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#RandomPositionsSamplerHiC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.samplers.online_sampler.OnlineSampler</span></code></p>
<p>This sampler randomly selects a region in the genome and retrieves
sequence and relevant Hi-C and optionally multibin genomic
data from that region. This implementation is modified based on
selene_sdk.samplers.RandomPositionSampler.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reference_sequence</strong> (<a class="reference external" href="https://selene.flatironinstitute.org/sequences.html#selene_sdk.sequences.Genome" title="(in Selene vmaster (0.4.8))"><em>selene_sdk.sequences.Genome</em></a>) – A genome to retrieve sequence from.</p></li>
<li><p><strong>target</strong> (<a class="reference internal" href="#selene_utils2.Genomic2DFeatures" title="selene_utils2.Genomic2DFeatures"><em>Genomic2DFeatures</em></a>) – Genomic2DFeatures object that loads the cooler files.</p></li>
<li><p><strong>features</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em>) – List of names that correspond to the cooler files.</p></li>
<li><p><strong>target_1d</strong> (<a class="reference internal" href="#selene_utils2.MultibinGenomicFeatures" title="selene_utils2.MultibinGenomicFeatures"><em>MultibinGenomicFeatures</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a><em>, </em><em>optional</em>) – MultibinGenomicFeatures object that loads 1D genomic feature data.</p></li>
<li><p><strong>background_cis_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a><em>, </em><em>optional</em>) – Path to the numpy file that stores the distance-based
expected background balanced scores for cis-interactions. If
specified with background_trans_file, the sampler will
return corresponding background array that matches with
the 2D feature retrieved.</p></li>
<li><p><strong>background_trans_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a><em>, </em><em>optional</em>) – Path to the numpy file that stores the expected background
balanced scores for trans-interactions. See doc for
<cite>background_cis_file</cite> for more detail.</p></li>
<li><p><strong>seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – Default is 436. Sets the random seed for sampling.</p></li>
<li><p><strong>validation_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em><em>, </em><em>optional</em>) – Default is <cite>[‘chr6’, ‘chr7’]</cite>. Holdout can be regional or
proportional. If regional, expects a list (e.g. <cite>[‘chrX’, ‘chrY’]</cite>).
Regions must match those specified in the first column of the
tabix-indexed BED file. If proportional, specify a percentage
between (0.0, 1.0). Typically 0.10 or 0.20.</p></li>
<li><p><strong>test_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>)</em><em>, </em><em>optional</em>) – Default is <cite>[‘chr8’, ‘chr9’]</cite>. See documentation for
<cite>validation_holdout</cite> for additional information.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – Default is 1000000. Model is trained on sequences of size
<cite>sequence_length</cite> where genomic features are retreived
for the same regions as the sequences.</p></li>
<li><p><strong>max_seg_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a><em>, </em><em>optional</em>) – Default is None. If specified and cross_chromosome is True,
bound the maximum length of each sequence segment.</p></li>
<li><p><strong>length_schedule</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>)</em><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a><em>, </em><em>optional</em>) – Default is None. If specified and cross_chromosome is True,
decide the sequence segment length to sample according to the
length schedule (before trimming to fit in the sequence length).
The length schedule is in the format of <cite>[p, [min_len, max_len]]</cite>,
which means, with probability <cite>p</cite>, decide the length by randomly
sampling an integer between <cite>min_len</cite> and <cite>max_len</cite>, and retrieve
the maximal remaining length as default with probability <cite>1-p</cite>.</p></li>
<li><p><strong>position_resolution</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – Default is 1. Preprocess the sampled start position by
<cite>start = start - start % position_resolution</cite>. Useful for binned
data.</p></li>
<li><p><strong>random_shift</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – Default is 0. Shift the coordinates to retrieve
sequence by a random integer in the range of [-random_shift, random_shift).</p></li>
<li><p><strong>random_strand</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – Default is True. If True, randomly select the strand of the
sequence, otherwise alway use the ‘+’ strand.</p></li>
<li><p><strong>cross_chromosome</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – Default is True. If True, allows sampling multiple segments of
sequences and the corresponding features. The default is sampling
the maximum length allowed by sequence_length, thus multiple segments
will only be sampled if <cite>sequence_length</cite> is larger than the minimum
chromosome length or when max_seg_length and length_schedule is specified
to limit the sequence segment length.</p></li>
<li><p><strong>permute_segments</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – Default is False. If True, permute the order of segments when
multiple segments are sampled.</p></li>
<li><p><strong>mode</strong> (<em>{'train'</em><em>, </em><em>'validate'</em><em>, </em><em>'test'}</em>) – Default is <cite>‘train’</cite>. The mode to run the sampler in.</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.reference_sequence">
<code class="sig-name descname">reference_sequence</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.reference_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>A genome to retrieve sequence from.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://selene.flatironinstitute.org/sequences.html#selene_sdk.sequences.Genome" title="(in Selene vmaster (0.4.8))">selene_sdk.sequences.Genome</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.target">
<code class="sig-name descname">target</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.target" title="Permalink to this definition">¶</a></dt>
<dd><p>The <cite>selene_sdk.targets.Target</cite> object holding the features that we
would like to predict.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://selene.flatironinstitute.org/targets.html#selene_sdk.targets.Target" title="(in Selene vmaster (0.4.8))">selene_sdk.targets.Target</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.target_1d">
<code class="sig-name descname">target_1d</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.target_1d" title="Permalink to this definition">¶</a></dt>
<dd><p>MultibinGenomicFeatures object that loads 1D genomic feature data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#selene_utils2.MultibinGenomicFeatures" title="selene_utils2.MultibinGenomicFeatures">MultibinGenomicFeatures</a> or <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a>, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.background_cis">
<code class="sig-name descname">background_cis</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.background_cis" title="Permalink to this definition">¶</a></dt>
<dd><p>One-dimensional numpy.ndarray that stores the distance-based
expected background balanced scores for cis-interactions.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.background_trans">
<code class="sig-name descname">background_trans</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.background_trans" title="Permalink to this definition">¶</a></dt>
<dd><p>The expected background balanced score for trans-interactions.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.bg">
<code class="sig-name descname">bg</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.bg" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether the sample will retrieve background arrays.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.validation_holdout">
<code class="sig-name descname">validation_holdout</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.validation_holdout" title="Permalink to this definition">¶</a></dt>
<dd><p>The samples to hold out for validating model performance. These
can be “regional” or “proportional”. If regional, this is a list
of region names (e.g. <cite>[‘chrX’, ‘chrY’]</cite>). These regions must
match those specified in the first column of the tabix-indexed
BED file. If proportional, this is the fraction of total samples
that will be held out.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.test_holdout">
<code class="sig-name descname">test_holdout</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.test_holdout" title="Permalink to this definition">¶</a></dt>
<dd><p>The samples to hold out for testing model performance. See the
documentation for <cite>validation_holdout</cite> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.sequence_length">
<code class="sig-name descname">sequence_length</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.sequence_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Model is trained on sequences of size
<cite>sequence_length</cite> where genomic features are retreived
for the same regions as the sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.max_seg_length">
<code class="sig-name descname">max_seg_length</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.max_seg_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Default is None. If specified and cross_chromosome is True,
bound the maximum length of each sequence segment.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a> or <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.length_schedule">
<code class="sig-name descname">length_schedule</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.length_schedule" title="Permalink to this definition">¶</a></dt>
<dd><p>Default is None. If specified and cross_chromosome is True,
decide the sequence segment length to sample according to the
length schedule (before trimming to fit in the sequence length).
The length schedule is in the format of <cite>[p, [min_len, max_len]]</cite>,
which means, with probability <cite>p</cite>, decide the length by randomly
sampling an integer between <cite>min_len</cite> and <cite>max_len</cite>, and retrieve
the maximal remaining length as default with probability <cite>1-p</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a>)) or <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)">None</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.position_resolution">
<code class="sig-name descname">position_resolution</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.position_resolution" title="Permalink to this definition">¶</a></dt>
<dd><p>Default is 1. Preprocess the sampled start position by
<cite>start = start - start % position_resolution</cite>. Useful for binned
data.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.random_shift">
<code class="sig-name descname">random_shift</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.random_shift" title="Permalink to this definition">¶</a></dt>
<dd><p>Default is 0. Shift the coordinates to retrieve
sequence by a random integer in the range of [-random_shift, random_shift).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.random_strand">
<code class="sig-name descname">random_strand</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.random_strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Default is True. If True, randomly select the strand of the
sequence, otherwise alway use the ‘+’ strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.cross_chromosome">
<code class="sig-name descname">cross_chromosome</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.cross_chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Default is True. If True, allows sampling multiple segments of
sequences and the corresponding features. The default is sampling
the maximum length allowed by sequence_length, thus multiple segments
will only be sampled if <cite>sequence_length</cite> is larger than the minimum
chromosome length or when max_seg_length and length_schedule is specified
to limit the sequence segment length.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.permute_segments">
<code class="sig-name descname">permute_segments</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.permute_segments" title="Permalink to this definition">¶</a></dt>
<dd><p>Default is False. If True, permute the order of segments when
multiple segments are sampled.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.modes">
<code class="sig-name descname">modes</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.modes" title="Permalink to this definition">¶</a></dt>
<dd><p>The list of modes that the sampler can be run in.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="selene_utils2.RandomPositionsSamplerHiC.mode">
<code class="sig-name descname">mode</code><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.mode" title="Permalink to this definition">¶</a></dt>
<dd><p>The current mode that the sampler is running in. Must be one of
the modes listed in <cite>modes</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="selene_utils2.RandomPositionsSamplerHiC.init">
<code class="sig-name descname">init</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#RandomPositionsSamplerHiC.init"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.init" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="selene_utils2.RandomPositionsSamplerHiC.sample">
<code class="sig-name descname">sample</code><span class="sig-paren">(</span><em class="sig-param">batch_size=1</em>, <em class="sig-param">mode=None</em>, <em class="sig-param">coordinate_only=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_utils2.html#RandomPositionsSamplerHiC.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_utils2.RandomPositionsSamplerHiC.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Randomly draws a mini-batch of examples and their corresponding
labels.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em>) – Default is 1. The number of examples to include in the
mini-batch.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a><em>, </em><em>optional</em>) – Default is None. The operating mode that the object should run in.
If None, will use the current mode <cite>self.mode</cite>.</p></li>
<li><p><strong>coordinate_only</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a><em>, </em><em>optional</em>) – Default is False. If True, only return the coordinates.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sequences, targets, …</strong> – A tuple containing the numeric representation of the
sequence examples, their corresponding 2D targets, and optionally 1D targets
(if target_1d were specified) and background
matrices (if background_cis_file and background_trans_file were
specified). The shape of <cite>sequences</cite> will be
<span class="math notranslate nohighlight">\(B \times L \times N\)</span>, where <span class="math notranslate nohighlight">\(B\)</span> is
<cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length, and
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet.
The shape of <cite>targets</cite> will be <span class="math notranslate nohighlight">\(B \times M \times M\)</span>,
where <span class="math notranslate nohighlight">\(M \times M\)</span> is target.shape. The shape of 1D targets
is <span class="math notranslate nohighlight">\(B \times K \times F\)</span>, where <span class="math notranslate nohighlight">\(K = \)</span>
and <span class="math notranslate nohighlight">\(F =\)</span> <cite>self.n_features</cite>. The shape of background matrices
are the same as <cite>targets</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)">tuple</a>(<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.20)">numpy.ndarray</a>, ..)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="selene_utils2.SampleIndices">
<em class="property">class </em><code class="sig-prename descclassname">selene_utils2.</code><code class="sig-name descname">SampleIndices</code><span class="sig-paren">(</span><em class="sig-param">indices</em>, <em class="sig-param">weights</em><span class="sig-paren">)</span><a class="headerlink" href="#selene_utils2.SampleIndices" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></a></p>
<dl class="method">
<dt id="selene_utils2.SampleIndices.indices">
<em class="property">property </em><code class="sig-name descname">indices</code><a class="headerlink" href="#selene_utils2.SampleIndices.indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="method">
<dt id="selene_utils2.SampleIndices.weights">
<em class="property">property </em><code class="sig-name descname">weights</code><a class="headerlink" href="#selene_utils2.SampleIndices.weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="orca_modules.html" class="btn btn-neutral float-left" title="orca_modules module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Zhou Laboratory.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>